adapters/avail/src/service.rs

```diff
@@ -0,0 +1,214 @@
+use subxt::tx::PairSigner;
+use subxt::OnlineClient;
+use tracing::info;
+
+use crate::avail::{Confidence, ExtrinsicsData};
+use crate::spec::block::AvailBlock;
+use crate::spec::header::AvailHeader;
+use crate::spec::transaction::AvailBlobTransaction;
+use crate::spec::DaLayerSpec;
+use crate::verifier::Verifier;
+
+/// Runtime configuration for the DA service
+#[derive(Clone, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct DaServiceConfig {
+    pub light_client_url: String,
+    pub node_client_url: String,
+    //TODO: Safer strategy to load seed so it is not accidentally revealed.
+    pub seed: String,
+}
+
+#[derive(Clone)]
+pub struct DaProvider {
+    pub node_client: OnlineClient<AvailConfig>,
+    pub light_client_url: String,
+    signer: PairSigner<AvailConfig, Pair>,
+}
+
+impl DaProvider {
+    fn appdata_url(&self, block_num: u64) -> String {
+        let light_client_url = self.light_client_url.clone();
+        format!("{light_client_url}/v1/appdata/{block_num}")
+    }
+
+    fn confidence_url(&self, block_num: u64) -> String {
+        let light_client_url = self.light_client_url.clone();
+        format!("{light_client_url}/v1/confidence/{block_num}")
+    }
+
+    pub async fn new(config: DaServiceConfig) -> Self {
+        let pair = Pair::from_string_with_seed(&config.seed, None).unwrap();
+        let signer = PairSigner::<AvailConfig, Pair>::new(pair.0.clone());
+
+        let node_client = avail_subxt::build_client(config.node_client_url.to_string(), false)
+            .await
+            .unwrap();
+        let light_client_url = config.light_client_url;
+
+        DaProvider {
+            node_client,
+            light_client_url,
+            signer,
+        }
+    }
+}
+
+const POLLING_TIMEOUT: Duration = Duration::from_secs(60);
+const POLLING_INTERVAL: Duration = Duration::from_secs(1);
+
+// TODO: Is there a way to avoid coupling to tokio?
+
+async fn wait_for_confidence(confidence_url: &str) -> anyhow::Result<()> {
+    let start_time = std::time::Instant::now();
+
+    loop {
+        if start_time.elapsed() >= POLLING_TIMEOUT {
+            return Err(anyhow!("Timeout..."));
+        }
+
+        let response = reqwest::get(confidence_url).await?;
+        if response.status() != StatusCode::OK {
+            info!("Confidence not received");
+            tokio::time::sleep(POLLING_INTERVAL).await;
+            continue;
+        }
+
+        let response: Confidence = serde_json::from_str(&response.text().await?)?;
+        if response.confidence < 92.5 {
+            info!("Confidence not reached");
+            tokio::time::sleep(POLLING_INTERVAL).await;
+            continue;
+        }
+
+        break;
+    }
+
+    Ok(())
+}
+
+async fn wait_for_appdata(appdata_url: &str, block: u32) -> anyhow::Result<ExtrinsicsData> {
+    let start_time = std::time::Instant::now();
+
+    loop {
+        if start_time.elapsed() >= POLLING_TIMEOUT {
+            return Err(anyhow!("Timeout..."));
```

Comment:
Could we please extend error message to include what was timeout value that was exceeded?